cmd/branchingconfigmanagers/bugzilla-config-manager/main.go

```diff
@@ -0,0 +1,197 @@
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to gather options")
+	}
+	configBytes, err := ioutil.ReadFile(opts.lifecycleConfigFile)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to read --lifecycle-config")
+	}
+
+	var lifecycleConfig ocplifecycle.Config
+	if err := yaml.Unmarshal(configBytes, &lifecycleConfig); err != nil {
+		logrus.WithError(err).Fatal("failed to deserialize the config")
+	}
+
+	configPath := path.Join(opts.pluginConfigDir, config.PluginConfigFile)
+	agent := plugins.ConfigAgent{}
+	if err := agent.Load(configPath, []string{opts.pluginConfigDir}, "_pluginconfig.yaml", false); err != nil {
+		logrus.WithError(err).Fatal("failed to load Prow plugin configuration")
+	}
+
+	pluginConfig := agent.Config()
+	updatedBugzillaConfig, err := run(lifecycleConfig, pluginConfig.Bugzilla)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to reconcile Bugzilla config")
+	}
+
+	pluginConfig.Bugzilla = *updatedBugzillaConfig
+
+	pluginConfig, err = prowconfigsharding.WriteShardedPluginConfig(pluginConfig, afero.NewBasePathFs(afero.NewOsFs(), opts.pluginConfigDir))
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to write plugin config shards")
+	}
+	pluginConfigRaw, err := yaml.Marshal(pluginConfig)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to serialize plugin config")
+	}
+
+	if err := ioutil.WriteFile(configPath, pluginConfigRaw, 0644); err != nil {
+		logrus.WithError(err).Fatal("failed to write main plugin config")
+	}
+}
+
+var releaseBranchPrefixes = []string{"openshift-", "release-"}
+var developmentDependentBugStates = &[]plugins.BugzillaBugState{{Status: "MODIFIED"}, {Status: "ON_QA"}, {Status: "VERIFIED"}}
+
+func run(lifecycleConfig ocplifecycle.Config, currentConfig plugins.Bugzilla) (*plugins.Bugzilla, error) {
+	developmentVersion, gaVersions, err := extractFromOracle(lifecycleConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract config from lifecycleConfig: %w", err)
+	}
+
+	return reconcileConfig(&currentConfig, developmentVersion, gaVersions), nil
+}
+
+func extractFromOracle(lifecycleConfig ocplifecycle.Config) (developmentVersion *ocplifecycle.MajorMinor, gaVersions []ocplifecycle.MajorMinor, err error) {
+	var errs []error
+	var developmentVersionFound bool
+
+	for ocpProduct, productConfig := range lifecycleConfig {
+		if ocpProduct != "ocp" {
+			continue
+		}
+
+		for productVersion, config := range productConfig {
+			if len(config.Previous) < 1 {
+				continue
+			}
+
+			// TODO: We assume the first event is the last - Is that reasonable?
+			switch config.Previous[0].Event {
+			case ocplifecycle.LifecycleEventGenerallyAvailable:
+				parsedVersion, err := ocplifecycle.ParseMajorMinor(productVersion)
+				if err != nil {
+					errs = append(errs, fmt.Errorf("failed to parse %s as majorMinor version: %w", productVersion, err))
+					continue
+				}
+				gaVersions = append(gaVersions, *parsedVersion)
+			case ocplifecycle.LifecycleEventOpen:
```

Comment:
The latest event describes the expected current state. If we go back in the events we will find an `GA,Open` for all released versions, but that is irrelevant. In general, I think the config managers should be able to reconcile based on the current desired state and not depend on having reconciled a prior desired state, that would make everything super fragile.

Reply:
I think I disagree - we know that after open but before GA, a certain set of rules apply. If we, like you say, only look at the latest level we need to know all possible levels between GA an open for a verison, in order to know what to do, right ... ?

Reply:
> I think I disagree - we know that after open but before GA, a certain set of rules apply. If we, like you say, only look at the latest level we need to know all possible levels between GA an open for a verison, in order to know what to do, right ... ?

Could you then please describe these rules in https://issues.redhat.com/browse/DPTP-2441? As of written, everything can be inferred from the current phase without knowledge of history and phases are linear so we always know what came before and what will come next

Reply:
Is that not already exactly what the card says?

Reply:
> Before the version X.Y is open for development, ...

> After version X.Y is open for development ...

> after version X.Y GA, ...

Reply:
> Before the version X.Y is open for development, ...

That case doesn't actually have a phase and @petr-muller said its best to ignore it: https://github.com/openshift/ci-tools/pull/2272#discussion_r693001776

> After version X.Y is open for development ...
> after version X.Y GA, ...

Yeah those two cases have rules but they are derived from "last state is {open, ga}" and they don't care about what the state before that one was. Can you please point out concretely in which case we care about the history and can not derive the current configuration from the last state?

Reply:
I had a conversation with Steve about this in the past, and in a sense he is right, it's a tradeoff. I personally think that a manager should simply either know exactly what phase we are in and what should hold  in that phase, or do nothing (if we are in a phase where we do not know what should hold, do nothing). 

Steve pointed out that this makes you vulnerable to changes: if a program adds a phase, we need to make all managers aware of that phase, even if they need need to enforce the criteria of the previous phase. So what we could do is, if you find yourself in a phase where you have no separate criteria, you'd work backwards until you find a phase that you know, and enforce that.  The expense is that the observable logic is more complicated and the simple idea of "in a phase x, enforce y" ceases to hold.

> That case doesn't actually have a phase and @petr-muller said its best to ignore it: #2272 (comment)

I think that's a misunderstanding - that definitely is a "phase". You'd see this as a release with no past event and an `open` event in the future: that's certainly actionable. What I meant by that feedback is that we should not infer the state of 4.X+1 from the lifecycle information of 4.X. The 4.X+1 release should have its own lifecycle record.

Reply:
Also idk what if we deploy this manager after feature freeze, so open is not the next phase? It doesn't start "doing anything" until GA? The logic of this particular manager iterating through the list to figure out the last applicable thing is not that complex and makes it resilient to when it starts running or if it somehow misses a level ...

Reply:
> Steve pointed out that this makes you vulnerable to changes: if a program adds a phase, we need to make all managers aware of that phase, even if they need need to enforce the criteria of the previous phase. So what we could do is, if you find yourself in a phase where you have no separate criteria, you'd work backwards until you find a phase that you know, and enforce that

But this is a potentially dangerous assumption, isn't it? We assume "unknown phase" means "enforce most recent known phase" but we can not say today if that is correct or not. So doing nothing or failing seem better options to me. If we know that phase X means "skip it and realize phase X-N" then we can explicitly add code for that.

> . What I meant by that feedback is that we should not infer the state of 4.X+1 from the lifecycle information of 4.X. The 4.X+1 release should have its own lifecycle record

Ah ok. So should that handling exist? The jira issue doesn't mention it

Reply:
> But this is a potentially dangerous assumption, isn't it? 

Yeah, that was exactly what I was thinking. But at the same time, I realize that I have actually described the acceptance criteria in https://issues.redhat.com/browse/DPTP-2441 exactly like Steve proposes:

> Before the version X.Y is open for development, ensure that branches {openshift,release}-X.Y are validated by default with the following criteria:

^^^ This actually needs to hold before the branch is open, *after* the branch is open, durihng feature freeze and also during code freeze. But I did not bother to describe that explicitly. So I think Steve has a point.

> Ah ok. So should that handling exist? The jira issue doesn't mention it
I think it's okay to not have it for now.  There is very little config that's actually *needed* to exist before we even open a branch for development. Our current docs make sure we prepare future config in advance to make future switches easier, and I described the acceptance criteria based on that. But in a world where future switches are automated, preparing config "for the future" may be useless.

Reply:
It just seems like if we're going to enforce this "know exactly what to do in every phase" logic it's fragile because:

1. most of our config doesn't actually work like this, we latch at decision points and hold until new decision points - most config managers are not doing something different in every event
2. the release team has talked about increase the number of feature freeze events - one for teams that depend on the rebase, one for the rebase, one for unrelated things .... why should we need to recompile every manager every time?
3. if managers do nothing when they don't know the exact phase, it means we a) cannot deploy new managers in that phase and b) cannot add new repos in that phase

IMHO if we're going to be writing managers at random points in time they need to know how to start enforcing state at the moment they are deployed, not the next time in the next release they have something to do. Similarly, if a new repo is added that matches the criteria for the manager, it needs to make that repo conformant immediately, not in the next phase it happens to know something about.

Reply:
Like I honestly don't see the alternative - re-compile every manager every time we change the event list? Really?

Reply:
> So for all states other than GA or end of life?
Yes.

Reply:
I think I am convinced